common/src/main/kotlin/streams/service/sink/strategy/CUDIngestionStrategy.kt

```diff
@@ -0,0 +1,263 @@
+        return when (ids.contains("_id")) {
+            true -> "MATCH ($identifier:${getLabelsAsString(labels)}) WHERE id($identifier) = event.$fullField$ID_KEY._id"
+            else -> "$keyword ($identifier:${getLabelsAsString(labels)} {${getNodeKeysAsString(keys = ids, prefix = "$fullField$ID_KEY")}})"
+        }
+    }
+
+    private fun buildNodeCreateStatement(labels: List<String>): String = """
+            |${StreamsUtils.UNWIND}
+            |CREATE (n:${getLabelsAsString(labels)})
+            |SET n = event.properties
+        """.trimMargin()
+
+    private fun buildRelCreateStatement(from: NodeRelMetadata, to: NodeRelMetadata,
+                                        rel_type: String): String = """
+            |${StreamsUtils.UNWIND}
+            |${buildNodeLookupByIds(ids = from.ids, labels = from.labels, identifier = "from", field = "from")}
+            |${buildNodeLookupByIds(ids = to.ids, labels = to.labels, identifier = "to", field = "to")}
+            |CREATE (from)-[r:${rel_type.quote()}]->(to)
+            |SET r = event.properties
+        """.trimMargin()
+
+    private fun buildNodeMergeStatement(labels: List<String>, ids: Set<String>): String = """
+            |${StreamsUtils.UNWIND}
+            |${buildNodeLookupByIds(keyword = "MERGE", ids = ids, labels = labels)}
+            |SET n += event.properties
+        """.trimMargin()
+
+    private fun buildRelMergeStatement(from: NodeRelMetadata, to: NodeRelMetadata,
+                                        rel_type: String): String = """
+            |${StreamsUtils.UNWIND}
+            |${buildNodeLookupByIds(ids = from.ids, labels = from.labels, identifier = "from", field = "from")}
+            |${buildNodeLookupByIds(ids = to.ids, labels = to.labels, identifier = "to", field = "to")}
+            |MERGE (from)-[r:${rel_type.quote()}]->(to)
+            |SET r += event.properties
+        """.trimMargin()
+
+    private fun buildNodeUpdateStatement(labels: List<String>, ids: Set<String>): String = """
+            |${StreamsUtils.UNWIND}
+            |${buildNodeLookupByIds(ids = ids, labels = labels)}
+            |SET n += event.properties
+        """.trimMargin()
+
+    private fun buildRelUpdateStatement(from: NodeRelMetadata, to: NodeRelMetadata,
+                                       rel_type: String): String = """
+            |${StreamsUtils.UNWIND}
+            |${buildNodeLookupByIds(ids = from.ids, labels = from.labels, identifier = "from", field = "from")}
+            |${buildNodeLookupByIds(ids = to.ids, labels = to.labels, identifier = "to", field = "to")}
+            |MATCH (from)-[r:${rel_type.quote()}]->(to)
+            |SET r += event.properties
+        """.trimMargin()
+
+    private fun buildDeleteStatement(labels: List<String>, ids: Set<String>, detach: Boolean): String = """
+            |${StreamsUtils.UNWIND}
+            |${buildNodeLookupByIds(ids = ids, labels = labels)}
+            |${if (detach) "DETACH " else ""}DELETE n
```

Comment:
Hold on don't you always have to detach?  This is maybe a behavior quesiton.

If you don't always DETACH then if a user tries to delete a connected node it will fail and then you have a special case to handle.

This is probably a fine point to document, but I think we should always DETACH DELETE the node to guarantee that what the user asked for (delete the node) actually happens.   Then put in the documentation that "If you do this, it implies you're deleting the incident relationships too" because it has to be that way.

Reply:
We can give to the user the power to choose if DETACH or not. Wdyt?

Into the documentation there is an explanation of what the `detach` property means:
https://github.com/neo4j-contrib/neo4j-streams/pull/224/files#diff-96dcb9477dd8de76ceb7eb3db007e3dcR376

Do you think we can improve the description?

Reply:
Making optional as you've done is a good choice, with default true.  I like it.  In terms of the description, I would just add a few words to summarize that detach means delete any incident relationships when you delete a node.   For Neo4j pros this is obvious, but some users will be lazy and not understand what "detach" means.  Link to the Cypher docs on detach delete so they can research further.